### PR TITLE
ENH: Added mmapmodule.c to Allow mmap objects to access the madvise() system call

### PR DESCRIPTION
**Pull request by a-deeb**
Adding python/cpython#6172 Allow mmap objects to access the madvise() system call , based on @seberg recommendation ,and adding it numpy separately based on issue #13172 
I added the `mmapmodule.c`  module in the numpy/numpy/core/src/common/ directory

### CODE REVIEW

File: numpy/core/src/common/mmapmodule.c [hunk 1]
```diff
@@ -0,0 +1,1596 @@
+/*
+ /  Author: Sam Rushing <rushing@nightmare.com>
```
Comment (charris) [inline numpy/core/src/common/mmapmodule.c]: Nit: replace `/` by `*`.

File: numpy/core/src/common/mmapmodule.c [hunk 2]
```diff
@@ -0,0 +1,1596 @@
...
+#ifndef MS_WINDOWS
+#define UNIX
+# ifdef HAVE_FCNTL_H
+#  include <fcntl.h>
+# endif /* HAVE_FCNTL_H */
+#endif
+
+#ifdef MS_WINDOWS
+#include <windows.h>
+static int
```
Comment (charris) [inline numpy/core/src/common/mmapmodule.c]: Blank line before.

**Comment by charris** [comment]
Lots of style nits, more than I commented on.  Take a look at `doc/C_STYLE_GUIDE.rst.txt`.

File: numpy/core/src/common/mmapmodule.c [hunk 3]
```diff
@@ -0,0 +1,1596 @@
...
+ /
+ /
+ / This version of mmapmodule.c has been changed significantly
+ / from the original mmapfile.c on which it was based.
+ / The original version of mmapfile is maintained by Sam at
+ / ftp://squirl.nightmare.com/pub/python/python-ext.
+ / and further modified by a-deeb
+*/
+
+#define PY_SSIZE_T_CLEAN
```
Comment (charris) [inline numpy/core/src/common/mmapmodule.c]: Is this placement safe?

**Comment by eric-wieser** [comment]
Can't we add just the madvise function, rather than cloning a builtin module?

**Comment by seberg** [comment]
Yeah, if we put this in for madvise, then we should limit it to the single feature we actually need.

**Comment by a-deeb** [comment]
In terms of the mmap_madvise_method it is the main function needed for this issue , so should I remove some other methods in this file?

**Comment by a-deeb** [comment]
Should I remove methods that are not needed ? or is it fine the way it currently is.
Thanks again !

File: numpy/core/src/common/mmapmodule.c [hunk 4]
```diff
@@ -0,0 +1,1586 @@
+/*
+ * Author: Sam Rushing <rushing@nightmare.com>
+ *  
+ * Contributors: ZackerySpytz, Yotam Medini, a-deeb
+ * Modified to support mmap with offset - to map a 'window' of a file
```
Comment (mattip) [inline numpy/core/src/common/mmapmodule.c]: What is the license of this file? Having author and contributor attribution with no copyright release is IMO problematic, but I am not a lawyer. Where did it come from? Also, do we allow github names as contributors?

Comment (eric-wieser) [reply numpy/core/src/common/mmapmodule.c]: I think it's a complete copy of the one in cpython

**Comment by a-deeb** [comment]
yes, its from cpython

File: numpy/core/src/common/mmapmodule.c (hunk 4 above)
Comment (mattip) [reply numpy/core/src/common/mmapmodule.c]: The original header in cpython is different. If it is copied from there it I think it should at least maintain the original header, probably with the addition of this sentence from the [CPython LICENSE](https://github.com/python/cpython/blob/3.7/LICENSE#L70) "Copyright (c) 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019 Python Software Foundation; All Rights Reserved". Looking around at other files in the repo, that would seem to be enough.

**Comment by a-deeb** [comment]
Hi everyone,  I would like to know if there is anything else I could do on this file regarding this issue, let me know, I'm really excited about helping out  😄

**Comment by seberg** [comment]
@a-deeb well, we did comment above that this should be a minimal port of functionality. We do not need to copy paste all of the memorymap implementation from cpython probably? So to move this along, would it be possible to steal only the parts that we would actually have to put into numpy?

**Comment by a-deeb** [comment]
@seberg  so I should just have the mmap_madvise_method()  along with any of its dependent structs and methods and remove everything else? I'll get working that, Thanks again!

**Comment by eric-wieser** [comment]
I'm not really sure this approach is worth pursuing - it might be best just to wait for the upstream support, and to point users to the `ctypes` solution until then.

Alternatively, we could include the ctypes solution directly into `memmap.py`. I don't thing a C wrapper is going to be particularly useful here, especially not if the approach is to try and copy `mmap_module`, of which almost none is useful.

**Comment by seberg** [comment]
@a-deeb sorry, but I do not think we will be merging this with the whole of `mmap_module.c` if you can do a minimal version (maybe best using c-types to keep maintainability high), that would sound like it might go anyway. I appreciate the input, but just want to avoid you putting in a lot of effort in a direction which is not going to get merged in any case. If you do not have time/wish to explore the ctypes solution, it is also fine to close this. Thanks either way!

**Comment by seberg** [comment]
@a-deeb thanks for the PR, but I will close this for now. Would be happy about a PR implementing a more minimal approach to exposing this in `memmap`.

**Pull request closed without merging**